cyberark/conjur
# Optimize certificate cache

#### What does this PR do?
Adding cache mechanism for provider certificates that will refresh them from provider only when client fails to decode with cached certificate.

Fix cucumber tests - fix tests to lookup for correct errors in log

#### Any background context you want to provide?
OIDC authentication verifies given TokenID by retrieving certificate from OIDC provider and verifying token signature. Following this feature it won't retrieve certificate per each request but cache it. A new retrieval will be triggered only on failure.


#### What ticket does this PR close?
Connected to #927 

#### Where should the reviewer start?
read relevant section at :  https://ca-il-confluence.il.cyber-ark.com/display/rndp/Research++-+rails+caching+mechanism+and+cert+optimization#?lucidIFH-viewer-599ae143=1

Then start with certificate_cache.rb - it stands on its own.

#### How should this be manually tested?
- Run several oidc requests and verify only one time provider is accessed to retrieve certificate.
- Rotate provider certificate and verify OIDC handler knows how to recover and retrieve new certificate.
